src/quicr_client_raw_session.cpp

```diff
@@ -37,184 +39,240 @@ namespace quicr {
+  connect();
+}
 
-      messages::MessageBuffer msg_buffer{ data.value() };
+QuicRClientRawSession::QuicRClientRawSession(
+  std::shared_ptr<qtransport::ITransport> transport_in,
+  qtransport::LogHandler& logger)
+  : has_shared_transport{ true }
+  , log_handler(logger)
+  , transport(transport_in)
+{
+}
 
-      try {
-        client.handle(std::move(msg_buffer));
-      } catch (const messages::MessageBuffer::ReadException& e) {
-        client.log_handler.log(qtransport::LogLevel::info,
-                               "Dropping malformed message: " +
-                                 std::string(e.what()));
-        return;
-      } catch (const std::exception& e) {
-        client.log_handler.log(qtransport::LogLevel::info,
-                               "Dropping malformed message: " +
-                                 std::string(e.what()));
-        return;
-      } catch (...) {
-        client.log_handler.log(
-          qtransport::LogLevel::fatal,
-          "Received malformed message with unknown fatal error");
-        throw;
-      }
-    }
+QuicRClientRawSession::~QuicRClientRawSession()
+{
+  if (!has_shared_transport &&
+      transport->status() != qtransport::TransportStatus::Disconnected) {
+    disconnect();
   }
+}
 
-private:
-  QuicRClientRawSession& client;
-};
-
-void
-QuicRClientRawSession::make_transport(RelayInfo& relay_info,
-                                      qtransport::TransportConfig tconfig,
-                                      qtransport::LogHandler& logger)
+bool
+QuicRClientRawSession::connect()
 {
-  qtransport::TransportRemote server = {
-    .host_or_ip = relay_info.hostname,
-    .port = relay_info.port,
-    .proto = relay_info.proto == RelayInfo::Protocol::UDP
-               ? qtransport::TransportProtocol::UDP
-               : qtransport::TransportProtocol::QUIC,
-  };
-  transport_delegate = std::make_unique<QuicRTransportDelegate>(*this);
-
-  transport = qtransport::ITransport::make_client_transport(
-    server, std::move(tconfig), *transport_delegate, logger);
-
   transport_context_id = transport->start();
+  log_handler.log(qtransport::LogLevel::info,
+                  (std::ostringstream()
+                   << "Connecting session " << transport_context_id << "...")
+                    .str());
 
   client_status = ClientStatus::CONNECTING;
-
-  while (transport->status() == qtransport::TransportStatus::Connecting) {
-    std::ostringstream log_msg;
-    log_msg << "Waiting for client to be ready, got: "
-            << int(transport->status());
-    logger.log(qtransport::LogLevel::info, log_msg.str());
+  while (!stopping && client_status == ClientStatus::CONNECTING) {
     std::this_thread::sleep_for(std::chrono::milliseconds(100));
   }
 
-  client_status = ClientStatus::READY;
+  if (stopping) {
+    log_handler.log(qtransport::LogLevel::info,
+                    (std::ostringstream() << "Cancelling connecting session "
```

Comment:
Though we try hard, picoquic (and even harder, UDP) decide that they don't want to be stopped, and continue on without a care. So fully stopping this when we want to leave a call if we didn't connect is a bit of an issue.

Reply:
Created https://github.com/Quicr/libquicr/issues/57 for this.